risc0/zkvm/src/host/client/prove/mod.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright 2023 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub(crate) mod bonsai;
+pub(crate) mod external;
+#[cfg(feature = "prove")]
+pub(crate) mod local;
+
+use std::{path::PathBuf, rc::Rc};
+
+use anyhow::Result;
+use risc0_binfmt::{MemoryImage, Program};
+use risc0_zkvm_platform::{memory::MEM_SIZE, PAGE_SIZE};
+
+use self::{bonsai::BonsaiProver, external::ExternalProver};
+use crate::{is_dev_mode, ExecutorEnv, Receipt, VerifierContext};
+
+/// A Prover can execute a given [MemoryImage] and produce a [Receipt] that can
+/// be used to verify correct computation.
+pub trait Prover {
+    /// Return a name for this [Prover].
+    fn get_name(&self) -> String;
+
+    /// Prove the specified [MemoryImage].
+    fn prove(
+        &self,
+        env: ExecutorEnv<'_>,
+        ctx: &VerifierContext,
```

Comment:
It seems like maybe this should be removed from this method since `VerifierContext` cannot be serialized, and so can't be sent to a remote prover. It seems like this would be set when constructing the prover, and the `opts` here acts as the way for client to indicate which hash suite should be used.

Reply:
However, if we want to call verify within prove on the client side, we'd still need it. I have a new branch which might make this obsolete, but this is a good point regardless.